### PR TITLE
JFG-879, tests for JaaS sessions, initial commit.

### PR DESCRIPTION
**Pull request by L-Evg**
**NOT** to be actually merged. For **preliminary** review purpose only.

### CODE REVIEW

File: test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/invoker/Invoker_Http.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+import java.net.URISyntaxException;
+import java.security.cert.X509Certificate;
+
+/**
+ * Honestly and boldly stolen from D. L., many thanks.
+ * Extremely simplified.
+ * //TODO: To be replaced by built-in Invoker implementation once Jggr2.0 is in the wild.
+ * //TODO: Do we really need it at the moment?
+ */
+public class Invoker_Http implements Invoker<String, HttpResponse, String> {
```
Comment (dlatnikov) [inline test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/invoker/Invoker_Http.java]: are you using it?

File: test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/validator/SessionsListResponseContentValidator.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        //Checks.
+        try {
+            int actlSize = actualSessions.size();
+            int expctdSize = TestContext.getSessions().size();
+            Assert.assertTrue("Several session records are expected. Check returned list's size", 1 < actlSize);
+            List<SessionEntity> noDuplicatesActualList = actualSessions.stream().distinct().collect(Collectors.toList());
+            Assert.assertEquals("Response contains duplicate session records", actlSize, noDuplicatesActualList.size());
+            //TODO: Wait for JFG-908 to be resolved and un-comment/re-factor.
+ //           Assert.assertTrue(String.format("Actual list(%d) is longer than expected one(%d).", actlSize, expctdSize), actlSize <= expctdSize);
+            // Re-factor one JFG-908 is resolved.
```
Comment (dlatnikov) [inline test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/validator/SessionsListResponseContentValidator.java]: one => once

File: test/jaas-test/src/main/resources/suite/defaults.conf.xml [hunk 3]
```diff
@@ -0,0 +1,66 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
```
Comment (dlatnikov) [inline test/jaas-test/src/main/resources/suite/defaults.conf.xml]: propose to remove all declaimers. if necessary we can auto generate them in the whole jagger project

File: test/jaas-test/src/main/resources/suite/defaults.conf.xml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    </scenario>
+
+    <!--This xml file will contain elements, shared between different configuration files-->
+    <!--Put here elements that will serve as parents for other elements -->
+
+    <!--Default test description - will serve as base for other test descriptions-->
+    <test-description id="td_default" version="1">
+
+        <!--Result collectors: put here validators, metrics, invokation listeners-->
+        <info-collectors>
```
Comment (dlatnikov) [inline test/jaas-test/src/main/resources/suite/defaults.conf.xml]: do we need these validator and metrics

File: test/jaas-test/src/main/resources/suite/defaults.conf.xml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        <endpoint-provider xsi:type="endpoint-provider-list">
+            <endpoint>${jaas.endpoint}</endpoint>
+        </endpoint-provider>
+    </scenario>
+
+    <!--This xml file will contain elements, shared between different configuration files-->
+    <!--Put here elements that will serve as parents for other elements -->
+
+    <!--Default test description - will serve as base for other test descriptions-->
+    <test-description id="td_default" version="1">
```
Comment (dlatnikov) [inline test/jaas-test/src/main/resources/suite/defaults.conf.xml]: looks like this test description is not used at all

File: test/jaas-test/src/main/resources/suite/suite.conf.xml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    <configuration id="JaaS_TestConfiguration">
+        <latency-percentiles>
+            <percentile>99</percentile>
+        </latency-percentiles>
+
+        <test-suite id="ts_JaaSTestSuit">
+            <test-group id="tg_JaaS_GET_Sessions">
+                <test id="t_JaaS_GET_Sessions_SR_eq._1.0" testDescription="td_JaaS_SessionsList">
+                    <load xsi:type="load-user-groups">
+                        <user count="2" startCount="2" startIn="0" startBy="0" life="5s" id="tl_JaaS_sessions" />
```
Comment (dlatnikov) [inline test/jaas-test/src/main/resources/suite/suite.conf.xml]: I think life should be longer than maxDuration of the test

File: test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/util/TestContext.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        }
+        return localInstance;
+    }
+
+    public static Set<SessionEntity> getSessions() {
+        return get().sessions;
+    }
+
+    /**
+     * Returns NULL is no session entity found.
```
Comment (dlatnikov) [inline test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/util/TestContext.java]: is -> if?

File: test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/validator/SessionResponseContentValidator.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Override
+    public String getName() {
+        return "SessionResponseContentValidator";
+    }
+
+    @Override
+    public boolean validate(HttpRequestBase query, E endpoint, HttpResponse result, long duration)  {
+        String content = result.getBody();
+        boolean isValid = true;
```
Comment (dlatnikov) [inline test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/validator/SessionResponseContentValidator.java]: propose to start from negative and set it true only when all checks are pass

File: test/jaas-test/src/main/resources/profiles/basic/default.environment.properties [hunk 9]
```diff
@@ -0,0 +1,110 @@
+# ****************************************************************** #
```
Comment (dlatnikov) [inline test/jaas-test/src/main/resources/profiles/basic/default.environment.properties]: you, don't need to change this file. environment properties are enough

File: test/jaas-test/src/main/resources/suite/collectors.conf.xml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dlatnikov) [inline test/jaas-test/src/main/resources/suite/collectors.conf.xml]: you can skip this file if it is not used

File: test/jaas-test/src/main/resources/suite/invokers.conf.xml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dlatnikov) [inline test/jaas-test/src/main/resources/suite/invokers.conf.xml]: is it used?

File: test/jaas-test/src/main/resources/suite/suite.conf.xml [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    <configuration id="JaaS_TestConfiguration">
+        <latency-percentiles>
+            <percentile>99</percentile>
+        </latency-percentiles>
+
+        <test-suite id="ts_JaaSTestSuit">
+            <test-group id="tg_JaaS_GET_Sessions">
+                <test id="t_JaaS_GET_Sessions_List_SR_eq._1.0" testDescription="td_JaaS_SessionsList">
+                    <load xsi:type="load-user-groups">
+                        <user count="2" startCount="2" startIn="0" startBy="0" life="12s" id="tl_JaaS_sessions" />
```
Comment (dlatnikov) [inline test/jaas-test/src/main/resources/suite/suite.conf.xml]: good practice to parametrize life and maxDuration via properties. xml parser can handle it
in this case you can change duration of the test without recompiling

File: test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/provider/QueryProvider_SessionIds.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * Provides queries like /jaas/sessions/{id}.
+ */
+public class QueryProvider_SessionIds extends QueryProvider_SessionsList {
+
+    public QueryProvider_SessionIds() {}
+
+    @Override
+    public Iterator iterator() {
+        queries.addAll(TestContext.getSessions().stream().map(s -> new HttpGet(uri + "/" + s.getId())).collect(Collectors.toList()));
```
Comment (L-Evg) [inline test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/provider/QueryProvider_SessionIds.java]: Ok, got it.

File: test/jaas-test/src/main/resources/profiles/basic/default.environment.properties (hunk 9 above)
Comment (L-Evg) [inline test/jaas-test/src/main/resources/profiles/basic/default.environment.properties]: Ok.

File: test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/validator/SessionResponseContentValidator.java (hunk 8 above)
Comment (L-Evg) [inline test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/validator/SessionResponseContentValidator.java]: Ok.

File: test/jaas-test/src/main/resources/suite/collectors.conf.xml (hunk 10 above)
Comment (L-Evg) [inline test/jaas-test/src/main/resources/suite/collectors.conf.xml]: Deleted.

File: test/jaas-test/src/main/resources/suite/invokers.conf.xml (hunk 11 above)
Comment (L-Evg) [inline test/jaas-test/src/main/resources/suite/invokers.conf.xml]: No, deleted.

File: test/jaas-test/src/main/resources/suite/suite.conf.xml (hunk 6 above)
Comment (L-Evg) [inline test/jaas-test/src/main/resources/suite/suite.conf.xml]: Corrected.

File: test/jaas-test/src/main/resources/suite/defaults.conf.xml (hunk 4 above)
Comment (L-Evg) [inline test/jaas-test/src/main/resources/suite/defaults.conf.xml]: Removed the whole td_default.

File: test/jaas-test/src/main/resources/suite/defaults.conf.xml (hunk 5 above)
Comment (L-Evg) [inline test/jaas-test/src/main/resources/suite/defaults.conf.xml]: Removed.

**Comment by dlatnikov** [comment]
👍

File: test/jaas-test/src/main/resources/suite/suite.conf.xml (hunk 12 above)
Comment (L-Evg) [inline test/jaas-test/src/main/resources/suite/suite.conf.xml]: Ok.

File: test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/invoker/Invoker_Http.java (hunk 1 above)
Comment (L-Evg) [inline test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/invoker/Invoker_Http.java]: Deleted.

File: test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/validator/SessionsListResponseContentValidator.java (hunk 2 above)
Comment (L-Evg) [inline test/jaas-test/src/main/java/com/griddynamics/jagger/test/jaas/validator/SessionsListResponseContentValidator.java]: Corrected.

File: test/jaas-test/src/main/resources/suite/defaults.conf.xml (hunk 3 above)
Comment (L-Evg) [inline test/jaas-test/src/main/resources/suite/defaults.conf.xml]: Done.

**Pull request merged**